src/FeedForwardNeuralNetwork.cpp

```diff
@@ -137,12 +137,34 @@ void FeedForwardNeuralNetwork::setBeta(const double * beta)
 
 void FeedForwardNeuralNetwork::randomizeBetas()
 {
-    std::random_device rdev;
-    std::mt19937_64 rgen = std::mt19937_64(rdev());
-    std::uniform_real_distribution<double> rd = std::uniform_real_distribution<double>(-3.,3.);
+    using namespace std;
+
+    random_device rdev;
+    mt19937_64 rgen = std::mt19937_64(rdev());
+    uniform_real_distribution<double> rd;
+
+    int nsource;
+    double bah;
+
+    for (vector<NNLayer *>::size_type i=0; i<_L.size(); ++i)
+    {
+        for (int j=0; j<_L[i]->getNUnits(); ++j)
+        {
+            if (_L[i]->getUnit(j)->getFeeder())
+            {
+                nsource = _L[i]->getUnit(j)->getFeeder()->getNBeta();
+                // target sigma to keep sum of weighted inputs in range [-4,4], assuming uniform distribution
+                // sigma = 8/sqrt(12) = (b-a)/sqrt(12) * m^(1/2)
+                bah = 4 * pow(nsource, -0.5); // (b-a)/2
+                rd = uniform_real_distribution<double>(-bah,bah);
 
-    // set betas to new random values
-    for (int i=0; i<this->getNBeta(); ++i) this->setBeta(i, rd(rgen));
+                for (int k=0; k<nsource; ++k)
+                {
+                    _L[i]->getUnit(j)->getFeeder()->setBeta(k, rd(rgen));
+                }
+            }
+        }
+    }
```

Comment:
I suggest that in the Activation Function implementation classes, these bounds should be specified and used.
Also, as you already pointed out, all the activation functions should then have at least mean 0, and possibly variance 1.
I create a Trello card for this.

Reply:
Yes, I am aware of this. But creating a general solution was beyond the scope of this card.